File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.ans = 0

    def maxAncestorDiff(self, root: TreeNode) -> int:
        if root is None:
            return 0
        self.dfs(root, root.val, root.val)
        return self.ans

    def dfs(self, root: TreeNode, curMax: int, curMin: int) -> None:
        if root is None:
            return
        self.ans = max(self.ans, max(abs(root.val - curMax), abs(root.val - curMin)))
        curMax = max(curMax, root.val)
        curMin = min(curMin, root.val)
        self.dfs(root.left, curMax, curMin)
        self.dfs(root.right, curMax, curMin)
        

File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans = 0;
    int maxAncestorDiff(TreeNode* root) {
        if(!root) return ans;
        dfs(root, root->val, root->val);
        return ans;
    }
    
    void dfs(TreeNode* root, int curMax, int curMin) {
        if(!root) return;
        ans = max(ans, max(abs(curMax-root->val), abs(curMin-root->val)));
        curMax = max(root->val, curMax);
        curMin = min(root->val, curMin);
        dfs(root->left, curMax, curMin);
        dfs(root->right, curMax, curMin);
    }
};